README.md
# Chroma - Shopify fork

Chroma is a color manipulation and palette generation library. It is heavily
inspired by and a very close Ruby port of the
[tinycolor.js](https://bgrins.github.io/TinyColor/)
library. Many thanks to [Brian Grinstead](http://www.briangrinstead.com/blog/)
for his hard work on that library.

Please don't hesitate to examine the code and make issues, feature requests,
or pull requests. Please refer to the [Contributing](#contributing) section
below.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'chroma'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install chroma

## Creating Colors

Colors are created via the `Chroma.paint` method. It expects any one of
many possible color formats as a string, including names, hexadecimal, rgb,
hsl, and hsv. As a convenience, a `String#paint` is also available for more
succinct color creation.

```ruby
# With Chroma.paint
Chroma.paint 'red'                       # named colors
Chroma.paint '#00ff00'                   # 6 character hexadecimal
Chroma.paint '#00f'                      # 3 character hexadecimal
Chroma.paint 'rgb(255, 255, 0)'          # rgb
Chroma.paint 'rgba(255, 255, 0, 0.5)'    # rgba
Chroma.paint 'hsl(60, 100%, 50%)'        # hsl with percentages
Chroma.paint 'hsl(60, 1, 0.5)'           # hsl with decimals
Chroma.paint 'hsla(60, 100%, 50%, 0.5)'  # hsla
Chroma.paint 'hsv(60, 100%, 50%)'        # hsv with percentages
Chroma.paint 'hsv(60, 1, 0.5)'           # hsv with decimals
Chroma.paint 'hsva(60, 100%, 50%, 0.75)' # hsva

# With String#paint
'red'.paint
'#00ff00'.paint
'#00f'.paint
'rgb(255, 255, 0)'.paint
'rgba(255, 255, 0, 0.5)'.paint
'hsl(60, 100%, 50%)'.paint
'hsla(60, 100%, 50%, 0.5)'.paint
'hsv(60, 100%, 50%)'.paint
'hsva(60, 100%, 50%. 0.5)'.paint
```

## Motivation

Chroma's major strength is manipulating colors and generating color palettes,
which allows you to easily generate dynamic colors, dynamic themes for a web
application, and more.

## Color Manipulation

#### Lighten

Lighten the color by a given amount. Defaults to 10.

```ruby
'red'.paint.lighten     #=> #ff3333
'red'.paint.lighten(20) #=> #ff6666
```

#### Brighten

Brighten the color by a given amount. Defaults to 10.

```ruby
'red'.paint.brighten     #=> #ff1a1a
'red'.paint.brighten(20) #=> #ff3333
```

#### Darken

Darken the color by a given amount. Defaults to 10.

```ruby
'red'.paint.darken     #=> #cc0000
'red'.paint.darken(20) #=> #990000
```

#### Desaturate

Desaturate the color by a given amount. Defaults to 10.

```ruby
'red'.paint.desaturate     #=> #f20d0d
'red'.paint.desaturate(20) #=> #e61919
```

#### Saturate

Saturate the color by a given amount. Defaults to 10.

```ruby
'#123'.paint.saturate     #=> #0e2236
'#123'.paint.saturate(20) #=> #0a223a
```

#### Grayscale

Convert the color to grayscale.

```ruby
'green'.paint.grayscale #=> #404040

# greyscale is an alias
'red'.paint.greyscale   #=> #808080
```

#### Spin

Spin a given amount in degrees around the hue wheel.

```ruby
'red'.paint.spin(30) #=> #ff8000
'red'.paint.spin(60) #=> yellow
'red'.paint.spin(90) #=> #80ff00
```

## Generating Palettes

Chroma's most powerful feature is palette generation. You can use the default
palettes available or even create your own custom palettes.

Palette methods are available via `Color#palette` and by default output an
array of colors. If you want the underlying color strings, you can pass in
the desired format via the `:as` option.

#### Available Formats

* name
* rgb
* hex
* hex6 (alias for hex)
* hex3
* hex8 (includes the alpha value in the highest order byte)
* hsl
* hsv

#### Complement

Generate a complement palette.

```ruby
'red'.paint.palette.complement            #=> [red, cyan]
'red'.paint.palette.complement(as: :name) #=> ['red', 'cyan']
'red'.paint.palette.complement(as: :hex)  #=> ['#ff0000', '#00ffff']
```

#### Triad

Generate a triad palette.

```ruby
'red'.paint.palette.triad            #=> [red, lime, blue]
'red'.paint.palette.triad(as: :name) #=> ['red', 'lime', 'blue']
'red'.paint.palette.triad(as: :hex)  #=> ['#ff0000', '#00ff00', '#0000ff']
```

#### Tetrad

Generate a tetrad palette.

```ruby
'red'.paint.palette.tetrad
#=> [red, #80ff00, cyan, #7f00ff]

'red'.paint.palette.tetrad(as: :name)
#=> ['red', '#80ff00', 'cyan', '#7f00ff']

'red'.paint.palette.tetrad(as: :hex)
#=> ['#ff0000', '#80ff00', '#00ffff', '#7f00ff']
```

#### Split Complement

Generate a split complement palette.

```ruby
'red'.paint.palette.split_complement
#=> [red, #ccff00, #0066ff]

'red'.paint.palette.split_complement(as: :name)
#=> ['red', '#ccff00', '#0066ff']

'red'.paint.palette.split_complement(as: :hex)
#=> ['#ff0000', '#ccff00', '#0066ff']
```

#### Analogous

Generate an analogous palette. Pass in a `:size` option to specify the size
of the palette (defaults to 6). Pass in a `:slice_by` option to specify the
angle size to slice into the hue wheel (defaults to 30 degrees).

```ruby
'red'.paint.palette.analogous
#=> [red, #ff0066, #ff0033, red, #ff3300, #ff6600]

'red'.paint.palette.analogous(as: :hex)
#=> ['#f00', '#f06', '#f03', '#f00', '#f30', '#f60']

'red'.paint.palette.analogous(size: 3)
#=> [red, #ff001a, #ff1a00]

'red'.paint.palette.analogous(size: 3, slice_by: 60)
#=> [red, #ff000d, #ff0d00]
```

#### Monochromatic

Generate a monochromatic palette. Pass in a `:size` option to specify the size
of the palette (defaults to 6).

```ruby
'red'.paint.palette.monochromatic
#=> [red, #2a0000, #550000, maroon, #aa0000, #d40000]

'red'.paint.palette.monochromatic(as: :hex)
#=> ['#ff0000', '#2a0000', '#550000', '#800000', '#aa0000', '#d40000']

'red'.paint.palette.monochromatic(size: 3)
#=> [red, #550000, #aa0000]
```

## Defining Custom Palettes

Chroma allows you to define your own custom palettes if the default ones aren't
all you're looking for. You can define a custom palette by calling
`Chroma.define_palette`, passing in a palette name and definition block. The
definition block uses the color manipulation methods (i.e. `lighten`, `spin`,
etc.) as its DSL. Every DSL call defines a new color that will be included
in the palette. Your seed color (i.e. the color from which you call the
palette method) will be included as the first color in your palette too.

```ruby
red = 'red'.paint

red.palette.respond_to? :my_palette #=> false

# Define a palette with 5 colors including the seed color
Chroma.define_palette :my_palette do
  spin 60
  spin 180
  spin(60).brighten(20) # chain calls as well
  greyscale
end

red.palette.respond_to? :my_palette #=> true

red.palette.my_palette #=> [#ff0000 #ffff00 #00ffff #ffff33 #808080]
```

## Dynamic Custom Palettes

You can generate custom palettes on the fly too with
`Chroma::Color#custom_palette`.

```ruby
'red'.paint.custom_palette do
  spin 60
  spin 180
end

#=> [red, yellow, cyan]
```

## Serializing Colors

Colors offer several methods to output to different string color [formats](#available-formats).

| Method    | Description                                                                                                      |
| --------- | ---------------------------------------------------------------------------------------------------------------- |
| `to_hsv`  | output to hsv string, outputs hsva if alpha < 1                                                                  |
| `to_hsl`  | output to hsl string, outputs hsla if alpha < 1                                                                  |
| `to_hex`  | output to hex string, optional argument allows 3-character hex output if possible                                |
| `to_hex8` | output to 8-character hex string with alpha value in the highest order byte                                      |
| `to_rgb`  | output to rgb string, outputs rgba if alpha < 1                                                                  |
| `to_name` | output to color name string if available, otherwise `'<unknown>'` or `to_hex` output based on optional arg value |
| `to_s`    | output to the appropriate string format based on how the color was created, optional arg forces the format       |

```ruby
# to_hsv
'red'.paint.to_hsv                  #=> 'hsv(0, 100%, 100%)'
'rgba(255, 0, 0, 0.5)'.paint.to_hsv #=> 'hsva(0, 100%, 100%, 0.5)'

# to_hsl
'red'.paint.to_hsl                  #=> 'hsl(0, 100%, 50%)'
'rgba(255, 0, 0, 0.5)'.paint.to_hsl #=> 'hsla(0, 100%, 50%, 0.5)'

# to_hex
'red'.paint.to_hex                  #=> '#ff0000'
'red'.paint.to_hex(true)            #=> '#f00'
'rgba(255, 0, 0, 0.5)'.paint.to_hex #=> '#ff0000'
'red'.paint.to_hex                  #=> '#ffff0000'
'rgba(255, 0, 0, 0.5)'.paint.to_hex #=> '#80ff0000'

# to_rgb
'red'.paint.to_rgb                  #=> 'rgb(255, 0, 0)'
'rgba(255, 0, 0, 0.5)'.paint.to_rgb #=> 'rgb(255, 0, 0, 0.5)'

# to_name
'red'.paint.to_name                  #=> 'red'
'#00f'.paint.to_name                 #=> 'blue'
'rgba(255, 0, 0, 0.5)'.paint.to_name #=> '<unknown>'
'#123'.paint.to_name(true)           #=> '#112233'

# to_s
'red'.paint.to_s             #=> 'red'
'rgb(255, 0, 0)'.paint.to_s  #=> 'rgb(255, 0, 0)'
'#f00'.paint.to_s            #=> '#f00'
'#80ff0000'.paint.to_s(:rgb) #=> 'rgba(255, 0, 0, 0.5)'
```

## Other Methods

Colors also have a few other helper methods:

| Method       | Description                                            |
| ------------ | ------------------------------------------------------ |
| `dark?`      | is the color dark?                                     |
| `light?`     | is the color light?                                    |
| `alpha`      | retrieve the alpha value                               |
| `brightness` | calculate the brightness as a number between 0 and 255 |
| `complement` | return the complementary color                         |

```ruby
# dark?
'red'.paint.dark?    #=> true
'yellow'.paint.dark? #=> false

# light?
'red'.paint.light?    #=> false
'yellow'.paint.light? #=> true

# alpha
'red'.paint.alpha                #=> 1.0
'rgba(0, 0, 0, 0.5)'.paint.alpha #=> 0.5

# brightness
'red'.paint.brightness    #=> 76.245
'yellow'.paint.brightness #=> 225.93
'white'.paint.brightness  #=> 255.0
'black'.paint.brightness  #=> 0.0

# complement
'red'.paint.complement #=> cyan
```

## Contributing

Please branch from **dev** for all pull requests.

1. Fork it (https://github.com/jfairbank/chroma/fork)
2. Checkout dev (`git checkout dev`)
3. Create your feature branch (`git checkout -b my-new-feature`)
4. Commit your changes (`git commit -am 'Add some feature'`)
5. Push to the branch (`git push origin my-new-feature`)
6. Create a new pull request against dev


CHANGELOG.md
## master

* Fix hex8 / #RGBA parsing and serialization.
* Add base error class (`Chroma::Error`).
* Change namespace for errors from `Chroma::Errors::` to `Chroma::`.
* Add red, green and blue color attributes
* Add hue, saturation, lightness color attributes

## v0.1.0 - 2016-05-26

### Dynamic Custom Palettes

You can generate custom palettes on the fly without predefining them now. The
old way of defining them with a name via `Chroma.define_palette` still works
too.

```ruby
# New dynamic way
'red'.paint.custom_palette do
  spin 60
  spin 180
end

#=> [red, yellow, cyan]
```

## v0.0.1 - 2015-01-14

**Method Changes:**

* Renamed options for analogous palette method.
  * `:results` -> `:size`
  * `:slices` -> `:slice_by`
* Renamed option for monochromatic palette method.
  * `:results` -> `:size`

**Miscellaneous Changes:**

* Add remaining specs for public API.
* Add "transparent" as color name for `Chroma.paint`.
* Minor API doc example fixes.
* Add public API usage examples to README.

## [v0.0.1.alpha.3] - 2015-01-13

**Bug Fixes:**

* Fix bug where `Color#complement` didn't return a color of the same format.
* Fix bug where palettes did not have the same format as the seed color. (#16)
* Fix bug where the helper method `bound01` was not always producing accurate
  results for percentages due to integer division.
* Fix bug where a `Color` created from an rgba string
  (e.g. `'rgba(255, 0, 0, 0.5).paint'`) was not serializing to an rgba string
  from `to_s`. (#17)

**Method Changes:**

* Add `Color#format` method to return `@format` instance variable.
* Change arguments for `analogous` and `monochromatic` to option arguments.
* Add ability to output palette as an array of color format strings via the
  `:as` option. (#10)
* On `Color` rename `greyscale` to `grayscale` and alias `greyscale` back
  to `grayscale`.

**Miscellaneous Changes:**

* Introduced custom errors and replaced `raise` calls with them.
* Added API doc headers. (#4)

## [v0.0.1.alpha.2] - 2015-01-13

**Bug Fixes:**

* Fixed bug with number of arguments passed to generator classes in RgbGenerator. (#1)
* Make `FromHexStringValues.from_hex8` take alpha as second parameter instead of last.
  Fixes incorrect color generation from hex8. (#6)
* Ensure that string serialization rounds the alpha value where applicable. (#7)
* Fix bug where `to_s` and `inspect` would return `<unknown>` instead of using
  hex if the format was `:name` and the named color could not be found. (#2)
* Fix bug where `Color` equality wasn't implemented. (#12)
* Fix bug where passing in an instance of `Hsl` or `Hsv` to `Color.new` caused
  their values to get changed. (#11)
* Fix bug with `Color` equality being off due to floating point math. (#13)
* Fix bug where `Color` instances generated from hsla and hsva strings had the
  wrong alpha value. (#15)

**Method Changes:**

* Add optional `hex_for_unknown` parameter to `Color::Serializers#to_name`.
  If true, it allows `to_name` to default to hex string if name is not found
  instead of returning `'<unknown>'`. (#2)
* Add missing conversion methods to converters (a12244f0d81c9480490cfb8a472993f54dd9fbd2)
* Add equality (`eql?` and `==`) methods to `Color` class and `ColorModes`
  classes. (#12, #13)
* Add `Chroma.define_palette` for defining custom palettes. (#9)
* Add `Color#paint` method for returning itself. (#14)
* Tweak `Color` serialization method names. Switched to this naming primarily
  to drop the `*_s` on the string serialization methods.
  * `to_hsv`    -> `hsv`
  * `to_hsv_s`  -> `to_hsv`
  * `to_hsl`    -> `hsl`
  * `to_hsl_s`  -> `to_hsl`
  * `to_hex`    -> `to_basic_hex` (made private)
  * `to_hex_s`  -> `to_hex`
  * `to_hex8`   -> `to_basic_hex8` (made private)
  * `to_hex8_s` -> `to_hex8`
  * `to_rgb`    -> `rgb` (moved attr_reader to serializers and made public)
  * `to_rgb_s`  -> `to_rgb`
  * Removed `to_name_s` alias

## [v0.0.1.alpha.1] - 2015-01-11

* Initial release


spec/chroma/paint_spec.rb
require 'spec_helper'

describe Chroma do
  describe '.paint' do
    context 'with named color' do
      it 'creates a color' do
        expect(Chroma.paint('red')).to be_a(Chroma::Color)
      end
    end

    context 'with 6 character hexadecimal' do
      it 'creates a color' do
        expect(Chroma.paint('#ff0000')).to be_a(Chroma::Color)
        expect(Chroma.paint('ff0000')).to be_a(Chroma::Color)
      end
    end

    context 'with 3 character hexadecimal' do
      it 'creates a color' do
        expect(Chroma.paint('#f00')).to be_a(Chroma::Color)
        expect(Chroma.paint('f00')).to be_a(Chroma::Color)
      end
    end

    context 'with 8 character hexadecimal' do
      let(:hex) { '#ff000080' }

      it 'creates a color' do
        expect(Chroma.paint(hex)).to be_a(Chroma::Color)
        expect(Chroma.paint(hex[1..-1])).to be_a(Chroma::Color)
      end

      it 'sets alpha' do
        expect(Chroma.paint(hex).alpha).to be_within(0.1).of(0.5)
      end
    end

    context 'with rgb' do
      it 'creates a color' do
        expect(Chroma.paint('rgb(255, 0, 0)')).to be_a(Chroma::Color)
        expect(Chroma.paint('rgba(255, 0, 0, 0.5)')).to be_a(Chroma::Color)
      end

      it 'sets alpha' do
        expect(Chroma.paint('rgba(255, 0, 0, 0.5)').alpha).to eq(0.5)
      end
    end

    context 'with hsl' do
      it 'creates a color' do
        expect(Chroma.paint('hsl(120, 100%, 50%)')).to be_a(Chroma::Color)
      end
    end
  end
end


spec/color/attributes_spec.rb
describe Chroma::Color do
  let(:red)    { 'red'.paint }
  let(:black)  { 'black'.paint }
  let(:white)  { 'white'.paint }
  let(:yellow) { 'yellow'.paint }

  describe '#dark?' do
    it 'returns true for dark colors' do
      expect(red).to   be_dark
      expect(black).to be_dark
    end

    it 'returns false for light colors' do
      expect(white).to_not  be_dark
      expect(yellow).to_not be_dark
    end
  end

  describe '#light?' do
    it 'returns false for dark colors' do
      expect(red).to_not   be_light
      expect(black).to_not be_light
    end

    it 'returns true for light colors' do
      expect(white).to  be_light
      expect(yellow).to be_light
    end
  end

  describe '#alpha' do
    it 'returns the correct alpha value' do
      expect('rgba(255, 0, 0, 0.75)'.paint.alpha).to eq 0.75
      expect('#ff000080'.paint.alpha).to             be_within(0.01).of(0.5)
      expect('transparent'.paint.alpha).to           eq 0
      expect('hsla(0, 100%, 50%, 0'.paint.alpha).to  eq 0
      expect(red.alpha).to                           eq 1
    end
  end

  describe '#red' do
    it 'returns the correct red value' do
      expect('rgba(255, 0, 0, 0.75)'.paint.red).to eq 255
      expect('#ff000080'.paint.red).to             eq 255
      expect('transparent'.paint.red).to           eq 0
      expect('hsla(0, 100%, 50%, 0'.paint.red).to  eq 255.0
      expect(red.red).to                           eq 255
    end
  end

  describe '#green' do
    it 'returns the correct green value' do
      expect('rgba(255, 128, 0, 0.75)'.paint.green).to eq 128
      expect('#ffaa0080'.paint.green).to               eq 170
      expect('transparent'.paint.green).to             eq 0
      expect('hsla(168, 100%, 50%, 0'.paint.green).to  eq 255.0
      expect(yellow.green).to                          eq 255
    end
  end

  describe '#blue' do
    it 'returns the correct blue value' do
      expect('rgba(255, 0, 75, 0.75)'.paint.blue).to eq 75
      expect('#ff00cc80'.paint.blue).to              eq 204
      expect('transparent'.paint.blue).to            eq 0
      expect('hsla(266, 100%, 25%, 0'.paint.blue).to eq 127.5
      expect(black.blue).to                          eq 0
    end
  end

  describe '#hue' do
    it 'returns the correct hue value' do
      expect('rgba(255, 0, 75, 0.75)'.paint.hue).to eq 342.3529411764706
      expect('#ff00cc80'.paint.hue).to              eq 312.0
      expect('transparent'.paint.hue).to            eq 0
      expect('hsla(266, 100%, 25%, 0'.paint.hue).to eq 266.0
      expect(red.hue).to                            eq 0
    end
  end

  describe '#saturation' do
    it 'returns the correct saturation value' do
      expect('rgba(212, 26, 75, 0.75)'.paint.saturation).to eq 78
      expect('#ff00cc80'.paint.saturation).to               eq 100
      expect('transparent'.paint.saturation).to             eq 0
      expect('hsla(266, 95%, 25%, 0'.paint.saturation).to   eq 95
      expect(black.saturation).to                           eq 0
    end
  end

  describe '#lightness' do
    it 'returns the correct lightness value' do
      expect('rgba(187, 30, 75, 0.75)'.paint.lightness).to eq 43
      expect('#ff00cc80'.paint.lightness).to               eq 50
      expect('transparent'.paint.lightness).to             eq 0
      expect('hsla(266, 100%, 25%, 0'.paint.lightness).to  eq 25
      expect(black.lightness).to                           eq 0
    end
  end

  describe '#brightness' do
    it 'returns the correct brightness' do
      expect(red.brightness).to    eq 76.245
      expect(black.brightness).to  eq 0
      expect(white.brightness).to  eq 255
      expect(yellow.brightness).to eq 225.93
    end
  end
end


spec/color/serializers_spec.rb
describe Chroma::Color do
  context 'with serializers' do
    let(:green)       { 'green'.paint }
    let(:blue)        { 'rgba(0, 0, 255, 0.5)'.paint }
    let(:transparent) { 'rgba(0, 0, 0, 0)'.paint }

    describe '#to_hsv' do
      it 'returns the hsv string' do
        expect(green.to_hsv).to eq('hsv(120, 100%, 50%)')
        expect(blue.to_hsv).to  eq('hsva(240, 100%, 100%, 0.5)')
      end
    end

    describe '#to_hsl' do
      it 'returns the hsl string' do
        expect(green.to_hsl).to eq('hsl(120, 100%, 25%)')
        expect(blue.to_hsl).to  eq('hsla(240, 100%, 50%, 0.5)')
      end
    end

    describe '#to_hex' do
      context 'with allow_3 set to false' do
        it 'returns the hex string' do
          expect(green.to_hex).to eq('#008000')
          expect(blue.to_hex).to  eq('#0000ff')
        end
      end

      context 'with allow_3 set to true' do
        it 'returns the hex string' do
          expect(green.to_hex(true)).to eq('#008000')
          expect(blue.to_hex(true)).to  eq('#00f')
        end
      end
    end

    describe '#to_hex8' do
      it 'returns the hex8 string' do
        expect(green.to_hex8).to eq('#008000ff')
        expect(blue.to_hex8).to  eq('#0000ff80')
      end
    end

    describe '#to_rgb' do
      it 'returns the rgb string' do
        expect(green.to_rgb).to eq('rgb(0, 128, 0)')
        expect(blue.to_rgb).to  eq('rgba(0, 0, 255, 0.5)')
      end
    end

    describe '#to_name' do
      context 'with hex_for_unknown set to false' do
        context 'with known named color' do
          context 'when alpha = 1' do
            it 'returns the named color' do
              expect(green.to_name).to eq('green')
            end
          end

          context 'when alpha < 1' do
            it 'returns "<unknown>"' do
              expect(blue.to_name).to eq('<unknown>')
            end
          end
        end

        context 'when alpha = 0' do
          it 'returns "transparent"' do
            expect(transparent.to_name).to eq('transparent')
          end
        end

        context 'with unknown named color' do
          it 'returns "<unknown>"' do
            expect('#123'.paint.to_name).to eq('<unknown>')
          end
        end
      end

      context 'with hex_for_unknown set to true' do
        context 'with known named color' do
          context 'when alpha = 1' do
            it 'returns the named color' do
              expect(green.to_name(true)).to eq('green')
            end
          end

          context 'when alpha < 1' do
            it 'returns the hex string' do
              expect(blue.to_name(true)).to eq('#0000ff')
            end
          end
        end

        context 'when alpha = 0' do
          it 'returns "transparent"' do
            expect(transparent.to_name(true)).to eq('transparent')
          end
        end

        context 'with unknown named color' do
          it 'returns returns the hex string' do
            expect('#123'.paint.to_name(true)).to eq('#112233')
          end
        end
      end
    end

    describe '#to_s' do
      it 'returns the appropriate string according to format' do
        expect('#ff0000'.paint.to_s).to                    eq('#ff0000')
        expect('#f00'.paint.to_s).to                       eq('#f00')
        expect('#80ff0000'.paint.to_s).to                  eq('#80ff0000')
        expect('hsl(120, 100%, 50%)'.paint.to_s).to        eq('hsl(120, 100%, 50%)')
        expect('hsla(120, 100%, 50%, 0.5)'.paint.to_s).to  eq('hsla(120, 100%, 50%, 0.5)')
        expect('hsv(120, 100%, 50%)'.paint.to_s).to        eq('hsv(120, 100%, 50%)')
        expect('hsva(120, 100%, 50%, 0.5)'.paint.to_s).to  eq('hsva(120, 100%, 50%, 0.5)')
        expect('red'.paint.to_s).to                        eq('red')
      end
    end

    describe '#hsv' do
      it 'returns an hsv instance' do
        hsv = green.hsv

        expect(hsv).to be_a(Chroma::ColorModes::Hsv)
        expect(hsv.h).to be_within(0.01).of(120)
        expect(hsv.s).to be_within(0.01).of(1)
        expect(hsv.v).to be_within(0.01).of(0.5)
        expect(hsv.a).to eq(1)
      end
    end

    describe '#hsl' do
      it 'returns an hsl instance' do
        hsl = green.hsl

        expect(hsl).to be_a(Chroma::ColorModes::Hsl)
        expect(hsl.h).to be_within(0.01).of(120)
        expect(hsl.s).to be_within(0.01).of(1)
        expect(hsl.l).to be_within(0.01).of(0.25)
        expect(hsl.a).to eq(1)
      end
    end

    describe '#rgb' do
      it 'returns the underlying @rgb iv' do
        expect(green.rgb).to equal(green.instance_variable_get(:@rgb))
      end
    end
  end
end


lib/chroma/color/attributes.rb
module Chroma
  class Color
    # Attribute methods for {Color}.
    module Attributes
      attr_reader :format

      # Determines if the color is dark.
      #
      # @example
      #   'red'.paint.dark?    #=> true
      #   'yellow'.paint.dark? #=> false
      #
      # @return [true, false]
      def dark?
        brightness < 128
      end

      # Determines if the color is light.
      #
      # @example
      #   'red'.paint.light?    #=> false
      #   'yellow'.paint.light? #=> true
      #
      # @return [true, false]
      def light?
        !dark?
      end

      # Returns the alpha channel value.
      #
      # @example
      #   'red'.paint.alpha                #=> 1.0
      #   'rgba(0, 0, 0, 0.5)'.paint.alpha #=> 0.5
      #
      # @return [Float]
      def alpha
        @rgb.a
      end

      def red
        @rgb.r
      end

      def green
        @rgb.g
      end

      def blue
        @rgb.b
      end

      def hue
        Chroma::Converters::HslConverter.convert_rgb(@rgb).h
      end

      def saturation
        (Chroma::Converters::HslConverter.convert_rgb(@rgb).s * 100).round
      end

      def lightness
        (Chroma::Converters::HslConverter.convert_rgb(@rgb).l * 100).round
      end

      # Calculates the brightness.
      #
      # @example
      #   'red'.paint.brightness    #=> 76.245
      #   'yellow'.paint.brightness #=> 225.93
      #
      # @return [Float]
      def brightness
        (@rgb.r * 299 + @rgb.g * 587 + @rgb.b * 114) / 1000.0
      end

      private

      def rounded_alpha
        @rounded_alpha ||= (alpha * 100).round / 100.0
      end
    end
  end
end


lib/chroma/color/serializers.rb
module Chroma
  class Color
    # Methods for serializing {Color} to different color mode string formats.
    module Serializers
      # Convert to hsv string.
      #
      # @example
      #   'red'.paint.to_hsv                  #=> 'hsv(0, 100%, 100%)'
      #   'rgba(255, 0, 0, 0.5)'.paint.to_hsv #=> 'hsva(0, 100%, 100%, 0.5)'
      #
      # @return [String]
      def to_hsv
        to_hs(:v)
      end

      # Convert to hsl string.
      #
      # @example
      #   'red'.paint.to_hsl                  #=> 'hsl(0, 100%, 50%)'
      #   'rgba(255, 0, 0, 0.5)'.paint.to_hsl #=> 'hsla(0, 100%, 50%, 0.5)'
      #
      # @return [String]
      def to_hsl
        to_hs(:l)
      end

      # Convert to hexadecimal string.
      #
      # @example
      #   'red'.paint.to_hex                  #=> '#ff0000'
      #   'red'.paint.to_hex(true)            #=> '#f00'
      #   'rgba(255, 0, 0, 0.5)'.paint.to_hex #=> '#ff0000'
      #
      # @param allow_3 [true, false] output 3-character hexadecimal
      #   if possible
      # @return [String]
      def to_hex(allow_3 = false)
        "##{to_basic_hex(allow_3)}"
      end

      # Convert to 8-character hexadecimal string. The highest order byte
      #   (left most hexadecimal pair represents the alpha value).
      #
      # @example
      #   'red'.paint.to_hex                  #=> '#ffff0000'
      #   'rgba(255, 0, 0, 0.5)'.paint.to_hex #=> '#80ff0000'
      #
      # @return [String]
      def to_hex8
        "##{to_basic_hex8}"
      end

      # Convert to rgb string.
      #
      # @example
      #   'red'.paint.to_rgb                  #=> 'rgb(255, 0, 0)'
      #   'rgba(255, 0, 0, 0.5)'.paint.to_rgb #=> 'rgb(255, 0, 0, 0.5)'
      #
      # @return [String]
      def to_rgb
        middle = @rgb.to_a[0..2].map(&:round).join(', ')

        with_alpha(:rgb, middle)
      end

      # Convert to named color if possible. If a color name can't be found, it
      # returns `'<unknown>'` or the hexadecimal string based on the value of
      # `hex_for_unknown`.
      #
      # @example
      #   'red'.paint.to_name                  #=> 'red'
      #   'rgba(255, 0, 0, 0.5)'.paint.to_name #=> '<unknown>'
      #   '#00f'.paint.to_name                 #=> 'blue'
      #   '#123'.paint.to_name(true)           #=> '#112233'
      #
      # @param hex_for_unknown [true, false] determine how unknown color names
      #   should be returned
      # @return [String]
      def to_name(hex_for_unknown = false)
        return 'transparent' if alpha.zero?

        if alpha < 1 || (name = Chroma.name_from_hex(to_basic_hex(true))).nil?
          if hex_for_unknown
            to_hex
          else
            '<unknown>'
          end
        else
          name
        end
      end

      # Convert to a string based on the color format.
      #
      # @example
      #   'red'.paint.to_s             #=> 'red'
      #   'rgb(255, 0, 0)'.paint.to_s  #=> 'rgb(255, 0, 0)'
      #   '#f00'.paint.to_s            #=> '#f00'
      #   '#80ff0000'.paint.to_s(:rgb) #=> 'rgba(255, 0, 0, 0.5)'
      #
      # @param format [Symbol] the color format
      # @return       [String]
      def to_s(format = @format)
        use_alpha = alpha < 1 && alpha >= 0 && /^hex(3|6)?$/ =~ format

        return to_rgb if use_alpha

        case format.to_s
        when 'rgb'         then to_rgb
        when 'hex', 'hex6' then to_hex
        when 'hex3'        then to_hex(true)
        when 'hex8'        then to_hex8
        when 'hsl'         then to_hsl
        when 'hsv'         then to_hsv
        when 'name'        then to_name(true)
        else                    to_hex
        end
      end

      alias_method :inspect, :to_s

      # Converts to an instance of {ColorModes::Hsv}
      # @return [ColorModes::Hsv]
      def hsv
        Converters::HsvConverter.convert_rgb(@rgb)
      end

      # Converts to an instance of {ColorModes::Hsl}
      # @return [ColorModes::Hsl]
      def hsl
        Converters::HslConverter.convert_rgb(@rgb)
      end

      # Converts to an instance of {ColorModes::Rgb}
      # @return [ColorModes::Rgb]
      attr_reader :rgb

      private

      def to_basic_hex(allow_3 = false)
        r, g, b = [@rgb.r, @rgb.g, @rgb.b].map do |n|
          to_2char_hex(n)
        end

        if allow_3 && r[0] == r[1] && g[0] == g[1] && b[0] == b[1]
          return "#{r[0]}#{g[0]}#{b[0]}"
        end

        "#{[r, g, b].flatten * ''}"
      end

      def to_basic_hex8
        [
          to_2char_hex(@rgb.r),
          to_2char_hex(@rgb.g),
          to_2char_hex(@rgb.b),
          to_2char_hex(alpha * 255)
        ].join('')
      end

      def to_hs(third)
        name = "hs#{third}"
        color = send(name)

        h = color.h.round
        s = (color.s * 100).round
        lv = (color.send(third) * 100).round

        middle = "#{h}, #{s}%, #{lv}%"

        with_alpha(name, middle)
      end

      def with_alpha(mode, middle)
        if alpha < 1
          "#{mode}a(#{middle}, #{rounded_alpha})"
        else
          "#{mode}(#{middle})"
        end
      end
    end
  end
end
